src/Datadog.Trace.DuckTyping/DuckType.Methods.cs

```diff
@@ -0,0 +1,605 @@
+                            targetMethod.GetParameters().Select(p => p.ParameterType).ToArray(),
+                            null);
+                    }
+
+                    ILHelpers.TypeConversion(dynIL, targetMethod.ReturnType, returnType);
+                    dynIL.Emit(OpCodes.Ret);
+
+                    // Emit the Call to the dynamic method pointer [Calli]
+                    il.Emit(OpCodes.Ldc_I8, (long)GetRuntimeHandle(dynMethod).GetFunctionPointer());
+                    il.Emit(OpCodes.Conv_I);
+                    il.EmitCalli(OpCodes.Calli, dynMethod.CallingConvention, dynMethod.ReturnType, dynParameters, null);
+                }
+
+                // We check if we have output or ref parameters to set in the proxy method
+                if (outputAndRefParameters != null)
+                {
+                    foreach (OutputAndRefParameterData outOrRefParameter in outputAndRefParameters)
+                    {
+                        Type proxyArgumentType = outOrRefParameter.ProxyArgumentType.GetElementType();
+                        Type localType = outOrRefParameter.LocalType.GetElementType();
+
+                        // We load the argument to be set
+                        ILHelpers.WriteLoadArgument(outOrRefParameter.ProxyArgumentIndex, il, false);
+
+                        // We load the value from the local
+                        ILHelpers.WriteLoadLocal(outOrRefParameter.LocalIndex, il);
+
+                        // If we detect duck chaining we create a new proxy instance with the output of the original target method
+                        if (!proxyArgumentType.IsValueType &&
+                            !proxyArgumentType.IsGenericParameter &&
+                            !proxyArgumentType.IsAssignableFrom(localType))
+                        {
+                            // We call DuckType.CreateCache<>.Create()
+                            MethodInfo getProxyMethodInfo = typeof(CreateCache<>)
+                                .MakeGenericType(proxyArgumentType).GetMethod("Create");
+
+                            il.Emit(OpCodes.Call, getProxyMethodInfo);
+                        }
+                        else
+                        {
+                            ILHelpers.TypeConversion(il, localType, proxyArgumentType);
+                        }
+
+                        // We store the value
+                        il.Emit(OpCodes.Stind_Ref);
+                    }
+                }
+
+                // Check if the target method returns something
+                if (targetMethod.ReturnType != typeof(void))
+                {
+                    // Handle the return value
+                    // Check if the type can be converted or if we need to enable duck chaining
+                    if (proxyMethodDefinition.ReturnType != targetMethod.ReturnType && !proxyMethodDefinition.ReturnType.IsValueType && !proxyMethodDefinition.ReturnType.IsAssignableFrom(targetMethod.ReturnType))
```

Comment:
This check seems very prevalent when checking a proxyType versus the realType. Do you think it's worth abstracting this into a Helper method?

Reply:
Yeah, I like the idea.

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/946/commits/e3020964dcbb2b80b94368777abbdd5b1de4e248